From 95ef285dcdd54e7a8cb7ad6b43b329479e51f61d Mon Sep 17 00:00:00 2001
From: liweigang <liweiganga@gmail.com>
Date: Tue, 9 Dec 2025 14:29:19 +0800
Subject: [PATCH] Fix strncpy buffer truncation warning in ini.c

- Replace strncpy(dest, src, size) with strncpy(dest, src, size - 1) in strncpy0()
  function to prevent GCC -Wstringop-truncation warning when destination size
  equals buffer size
- Ensure null-termination with dest[size - 1] = '\0' for safety
- This maintains backward compatibility while addressing compiler security warning
---
 src/lib/ini.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/lib/ini.c b/src/lib/ini.c
index f7db341..587de0d 100644
--- a/src/lib/ini.c
+++ b/src/lib/ini.c
@@ -90,8 +90,10 @@ static char* find_last_char_or_comment(const char* s, char c)
 /* Version of strncpy that ensures dest (size bytes) is null-terminated. */
 static char* strncpy0(char* dest, const char* src, size_t size)
 {
-    strncpy(dest, src, size);
-    dest[size - 1] = '\0';
+    if (size > 0) {
+        strncpy(dest, src, size - 1);
+        dest[size - 1] = '\0';
+    }
     return dest;
 }
 
